volume/mounts/mounts.go

```diff
@@ -127,16 +164,34 @@ func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun
 		}
 		volumePath, err := m.Volume.Mount(id)
 		if err != nil {
-			return "", errors.Wrapf(err, "error while mounting volume '%s'", m.Source)
+			return "", noCleanup, errors.Wrapf(err, "error while mounting volume '%s'", m.Source)
 		}
 
 		m.ID = id
+		clean := noCleanup
+		if m.Spec.VolumeOptions != nil && m.Spec.VolumeOptions.Subpath != "" {
+			subpath := m.Spec.VolumeOptions.Subpath
+
+			safePath, err := safepath.Join(volumePath, subpath)
+			if err != nil {
+				if err := m.Volume.Unmount(id); err != nil {
+					log.G(context.TODO()).WithError(err).Error("failed to unmount after safepath.Join failed")
+				}
+				return "", noCleanup, err
+			}
+			m.safePaths = append(m.safePaths, safePath)
```

Comment:
Why? What's wrong with letting users mount a volume to more than one mountpoint if they want to?

It may prove very tricky to get the check right on Windows (unless we use `os.SameFile` or similar) [because the Windows filesystem is not always case insensitive.](https://devblogs.microsoft.com/commandline/per-directory-case-sensitivity-and-wsl/)

Reply:
I don't think we need to restrict it.
It's safe for multiple `SafePath` to be created from the same `[base, subpath]` path.

On Linux each subpath gets its own independent temporary file/directory the target path gets bind mounted at so cleaning up one `SafePath` doesn't affect others.
On Windows, that will just be another set of open handles locking each part of the target path.

Reply:
Maybe I had my wires crossed and was thinking this was for the _destination_ (I'd have to look at the code again), if that's not at hand here, then ignore my comment.

My main concern here was that ending up with duplicate mounts can be troublesome, and we for sure had cases where those cases weren't detected properly (some due to matching happening case-sensitive); such cases have led to users scratching their heads why the "volume was empty" (because it was shadowed by another mount) etc. For such cases (where that being intentional would be "rare") I'd prefer to produce an error instead of silently accepting it (errors can be removed, but the revers (accepting now, and making it an error later) is much harder).

We do have some detection in place for those;

```bash
docker create -v hello:/foo -v world:/foo alpine
Error response from daemon: Duplicate mount point: /foo
```

Although not sure if our detection currently is complete (but possibly we're de-duplicating this case);

```bash
docker create -v /foo -v /foo alpine
552b828c29f596bdb0a0380de4ac1cd48bc9e93a6931540a437be618e82b399a
```